### DISCUSSION TITLE
How to create `Tensor` on GPU directly

### DISCUSSION THREAD

**Discussion by messense** [pykeio/ort#204] [Q&A]
Currently `Tensor::from_array` allocates the value on CPU, is there an equivalent method to allocate the value on GPU device directly?

**Comment by decahedron1** [comment]
[`Tensor::new`](https://docs.rs/ort/2.0.0-rc.2/ort/type.Tensor.html#method.new)

**Reply by messense** [reply]
But that API does not pass the data in?

**Reply by messense** [reply]
Given this example

```rust
let allocator = Allocator::new(
	&session,
	MemoryInfo::new(AllocationDevice::CUDAPinned, 0, AllocatorType::Device, MemoryType::CPUInput)?
)?;

let mut img_input = Tensor::<f32>::new(&allocator, [1, 128, 128, 3])?;
```

How to set the actual data for `img_input`?

**Reply by decahedron1** [reply]
ONNX Runtime doesn't provide APIs for that. You can grab the pointer to the data with [`Tensor::data_ptr_mut`](https://docs.rs/ort/2.0.0-rc.2/ort/type.Tensor.html#method.data_ptr), cast it to a more concrete data type (in the case of CUDA, this would be a CUDA buffer) and write the data that way.

**Reply by messense** [reply]
That's unfortunate, can it be done by modify `Tensor::from_array` and pass in a `memory_info` instead?

https://github.com/pykeio/ort/blob/ce5aaba2ca319bffb9dbdceb3fbc29d5dd013a7e/src/value/impl_tensor/create.rs#L180

**Reply by decahedron1** [reply]
The `memory_info` specifies where the data buffer itself resides, it won't copy data from CPU to GPU.

Maybe I/O binding is a better fit?

**Reply by messense** [reply]
Yeah I'm actually try to use I/O binding, but couldn't find an example that actually binds input/output.
It's a bit unclear how to use the I/O binding API correctly to reduce copies between CPU & GPU.
